File: app/src/main/java/com/hkb48/keepdo/NotificationController.java
package com.hkb48.keepdo;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.media.AudioAttributes;
import android.media.AudioManager;
import android.net.Uri;
import android.os.Build;
import android.util.Log;

import androidx.annotation.RequiresApi;
import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;

import com.hkb48.keepdo.settings.Settings;
import com.hkb48.keepdo.util.CompatUtil;

import java.util.List;

public class NotificationController {
    static final int NOTIFICATION_ID_HANDHELD = -1;
    private static final String TAG_KEEPDO = "#LOG_KEEPDO: ";
    private static final String GROUP_KEY_REMINDERS = "group_key_reminders";
    private static final String CHANNEL_ID = "Channel_ID";

    static void showReminder(final Context context, long taskId) {
        Task task = DatabaseAdapter.getInstance(context).getTask(taskId);
        String taskName = null;
        if (task != null) {
            taskName = task.getName();
            if (BuildConfig.DEBUG) {
                Log.v(TAG_KEEPDO, "taskId:" + taskId + ", taskName:" + taskName);
            }
        }

        NotificationCompat.Builder builder = new NotificationCompat.Builder(context, CHANNEL_ID);
        builder.setSmallIcon(R.drawable.ic_notification);
        builder.setTicker(taskName);
        builder.setDefaults(Notification.DEFAULT_LIGHTS);

        Settings.initialize(context.getApplicationContext());
        String reminderRingtone = Settings.getAlertsRingTone();
        if (reminderRingtone != null) {
            builder.setSound(Uri.parse(reminderRingtone));
        }
        String vibrateWhen = Settings.getAlertsVibrateWhen();
        boolean vibrateAlways = vibrateWhen.equals("always");
        boolean vibrateSilent = vibrateWhen.equals("silent");
        boolean nowSilent = false;
        AudioManager audioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
        if (audioManager != null) {
            nowSilent = audioManager.getRingerMode() == AudioManager.RINGER_MODE_VIBRATE;
        }

        // Possibly generate a vibration
        if (vibrateAlways || (vibrateSilent && nowSilent)) {
            builder.setDefaults(Notification.DEFAULT_VIBRATE);
        }
        Intent launchIntent = new Intent(context, TasksActivity.class);
        PendingIntent launchPendingIntent = PendingIntent.getActivity(context, 0, launchIntent, 0);
        builder.setContentIntent(launchPendingIntent);
        builder.setAutoCancel(true);

        Intent actionIntent = new Intent(context, ActionHandler.class);
        actionIntent.putExtra(ActionHandler.INTENT_EXTRA_TASK_ID, taskId);
        PendingIntent actionPendingIntent = PendingIntent.getService(context, (int) taskId, actionIntent, PendingIntent.FLAG_CANCEL_CURRENT);

        // Notification for hand-held device
        Bitmap largeIcon = BitmapFactory.decodeResource(context.getResources(),
                R.drawable.ic_launcher);
        builder.setLargeIcon(largeIcon);
        builder.setGroup(GROUP_KEY_REMINDERS);
        builder.setGroupSummary(true);

        ReminderManager reminderManager = ReminderManager.getInstance();
        List<Task> remainingTaskList = reminderManager.getRemainingUndoneTaskList(context);
        int numOfReminingTasks = remainingTaskList.size();

        if (numOfReminingTasks > 1) {
            NotificationCompat.InboxStyle inboxStyle = new NotificationCompat.InboxStyle();
            inboxStyle.setBigContentTitle(context.getString(R.string.notification_title_multi_tasks, numOfReminingTasks));
            for (Task remainingTask : remainingTaskList) {
                inboxStyle.addLine(remainingTask.getName());
            }

            builder.setContentTitle(context.getString(R.string.notification_title_multi_tasks, numOfReminingTasks));
            builder.setStyle(inboxStyle);
        } else {
            builder.setContentTitle(context.getString(R.string.notification_title_one_task));
            builder.setContentText(taskName);
            builder.addAction(R.drawable.ic_notification_check, context.getText(R.string.check_done), actionPendingIntent);
        }

        NotificationManagerCompat.from(context).notify(NOTIFICATION_ID_HANDHELD, builder.build());
    }

    @RequiresApi(Build.VERSION_CODES.O)
    static void createNotificationChannel(final Context context) {
        if (!CompatUtil.isNotificationChannelSupported()) {
            return;
        }

        // The user-visible name of the channel.
        CharSequence name = context.getString(R.string.channel_name);

        // The user-visible description of the channel.
        String description = context.getString(R.string.channel_description);

        int importance = NotificationManager.IMPORTANCE_DEFAULT;
        NotificationChannel mChannel = new NotificationChannel(CHANNEL_ID, name, importance);
        // Configure the notification channel.
        mChannel.setDescription(description);
        mChannel.enableLights(true);
        // Sets the notification light color for notifications posted to this
        // channel, if the device supports this feature.
        mChannel.setLightColor(Color.WHITE);
        Settings.initialize(context.getApplicationContext());
        String vibrateWhen = Settings.getAlertsVibrateWhen();
        boolean vibrateAlways = vibrateWhen.equals("always");
        boolean vibrateSilent = vibrateWhen.equals("silent");
        boolean nowSilent = false;
        AudioManager audioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
        if (audioManager != null) {
            nowSilent = audioManager.getRingerMode() == AudioManager.RINGER_MODE_VIBRATE;
        }

        // Possibly generate a vibration
        if (vibrateAlways || (vibrateSilent && nowSilent)) {
            mChannel.enableVibration(true);
            mChannel.setVibrationPattern(new long[]{100, 200, 300, 400, 500, 400, 300, 200, 400});
        }

        String reminderRingtone = Settings.getAlertsRingTone();
        if (reminderRingtone != null) {
            mChannel.setSound(Uri.parse(reminderRingtone), new AudioAttributes.Builder()
                    .setUsage(AudioAttributes.USAGE_NOTIFICATION)
                    .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC).build());
        }
        NotificationManager notificationManager =
                (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        if (notificationManager != null) {
            notificationManager.createNotificationChannel(mChannel);
        }
    }

    static void cancelReminder(final Context context) {
        NotificationManagerCompat.from(context).cancelAll();
    }

    static void cancelReminder(final Context context, long id) {
        int notificationId = (int) id;
        NotificationManagerCompat.from(context).cancel(notificationId);
    }

    @RequiresApi(Build.VERSION_CODES.O)
    public static String getNotificationChannelId() {
        return CHANNEL_ID;
    }
}

File: app/src/main/java/com/hkb48/keepdo/NotificationController.kt
package com.hkb48.keepdo

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.graphics.BitmapFactory
import android.graphics.Color
import android.media.AudioAttributes
import android.media.AudioManager
import android.net.Uri
import android.os.Build
import androidx.annotation.RequiresApi
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import com.hkb48.keepdo.db.entity.Task
import com.hkb48.keepdo.ui.TasksActivity
import com.hkb48.keepdo.ui.settings.Settings
import com.hkb48.keepdo.util.CompatUtil
import javax.inject.Inject

class NotificationController @Inject constructor(
    private val context: Context,
    private val reminderManager: ReminderManager
) {
    suspend fun showReminder(task: Task) {
        val taskName = task.name

        val builder = NotificationCompat.Builder(context, notificationChannelId)
            .setSmallIcon(R.drawable.ic_notification)
            .setTicker(taskName)
            .setDefaults(Notification.DEFAULT_LIGHTS)
        Settings.alertsRingTone?.let {
            builder.setSound(Uri.parse(it))
        }
        val vibrateWhen = Settings.alertsVibrateWhen
        val vibrateAlways = vibrateWhen == "always"
        val vibrateSilent = vibrateWhen == "silent"
        val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager?
        val nowSilent = audioManager?.ringerMode == AudioManager.RINGER_MODE_VIBRATE

        // Possibly generate a vibration
        if (vibrateAlways || vibrateSilent && nowSilent) {
            builder.setDefaults(Notification.DEFAULT_VIBRATE)
        }
        val launchIntent = Intent(context, TasksActivity::class.java)
        val launchPendingIntent = PendingIntent.getActivity(
            context, 0, launchIntent,
            CompatUtil.appendImmutableFlag(0)
        )
        builder.setContentIntent(launchPendingIntent)
        builder.setAutoCancel(true)

        // Notification for hand-held device
        val largeIcon = BitmapFactory.decodeResource(
            context.resources,
            R.drawable.ic_launcher
        )
        builder.setLargeIcon(largeIcon)
        builder.setGroup(GROUP_KEY_REMINDERS)
        builder.setGroupSummary(true)
        val remainingTaskList = reminderManager.getRemainingUndoneTaskList()
        val numOfRemainingTasks = remainingTaskList.size
        if (numOfRemainingTasks > 1) {
            val inboxStyle = NotificationCompat.InboxStyle()
            inboxStyle.setBigContentTitle(
                context.getString(
                    R.string.notification_title_multi_tasks,
                    numOfRemainingTasks
                )
            )
            for (remainingTask in remainingTaskList) {
                inboxStyle.addLine(remainingTask.name)
            }
            builder.setContentTitle(
                context.getString(
                    R.string.notification_title_multi_tasks,
                    numOfRemainingTasks
                )
            )
            builder.setStyle(inboxStyle)
        } else {
            val taskId = task._id
            val actionIntent = Intent(context, ActionReceiver::class.java).apply {
                putExtra(ActionReceiver.EXTRA_TASK_ID, taskId)
            }
            val actionPendingIntent = PendingIntent.getBroadcast(
                context,
                taskId,
                actionIntent,
                CompatUtil.appendImmutableFlag(PendingIntent.FLAG_CANCEL_CURRENT)
            )
            builder.setContentTitle(context.getString(R.string.notification_title_one_task))
            builder.setContentText(taskName)
            builder.addAction(
                R.drawable.ic_notification_check,
                context.getText(R.string.check_done),
                actionPendingIntent
            )
        }
        NotificationManagerCompat.from(context).notify(NOTIFICATION_ID_HANDHELD, builder.build())
    }

    @RequiresApi(Build.VERSION_CODES.O)
    fun createNotificationChannel() {
        if (CompatUtil.isNotificationChannelSupported().not()) {
            return
        }

        // The user-visible name of the channel.
        val name: CharSequence = context.getString(R.string.channel_name)

        // The user-visible description of the channel.
        val description = context.getString(R.string.channel_description)
        val importance = NotificationManager.IMPORTANCE_DEFAULT
        val channel = NotificationChannel(notificationChannelId, name, importance)
        // Configure the notification channel.
        channel.description = description
        channel.enableLights(true)
        // Sets the notification light color for notifications posted to this
        // channel, if the device supports this feature.
        channel.lightColor = Color.WHITE
        val vibrateWhen = Settings.alertsVibrateWhen
        val vibrateAlways = vibrateWhen == "always"
        val vibrateSilent = vibrateWhen == "silent"
        val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager?
        val nowSilent = audioManager?.ringerMode == AudioManager.RINGER_MODE_VIBRATE

        // Possibly generate a vibration
        if (vibrateAlways || vibrateSilent && nowSilent) {
            channel.enableVibration(true)
            channel.vibrationPattern = longArrayOf(100, 200, 300, 400, 500, 400, 300, 200, 400)
        }
        Settings.alertsRingTone?.let {
            channel.setSound(
                Uri.parse(it), AudioAttributes.Builder()
                    .setUsage(AudioAttributes.USAGE_NOTIFICATION)
                    .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC).build()
            )
        }
        val notificationManager =
            context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager?
        notificationManager?.createNotificationChannel(channel)
    }

    fun cancelReminder() {
        NotificationManagerCompat.from(context).cancelAll()
    }

    fun cancelReminder(id: Int) {
        NotificationManagerCompat.from(context).cancel(id)
    }

    companion object {
        const val NOTIFICATION_ID_HANDHELD = -1
        private const val GROUP_KEY_REMINDERS = "group_key_reminders"

        @get:RequiresApi(Build.VERSION_CODES.O)
        val notificationChannelId = "Channel_ID"
    }
}